File: backend/app/schemas/export.py
import re
from typing import List, Literal, Optional

from pydantic import BaseModel, Field, field_validator


class AccountFormPair(BaseModel):
    account_id: int = Field(..., gt=0, description="ID du compte Kobo a utiliser")
    form_uid: str = Field(..., min_length=1, description="UID du formulaire Kobo")


class ExportRequest(BaseModel):
    account_forms: List[AccountFormPair] = Field(..., min_length=1)
    form_name: str
    pivot_column: Optional[str] = None
    selected_columns: Optional[List[str]] = None
    selected_sheets: Optional[List[str]] = None
    filter_sites: Optional[List[str]] = None
    drive_folder_id: Optional[str] = None
    export_format: str = "xlsx"
    csv_separator: str = ";"
    csv_encoding: str = "utf-8-sig"
    csv_quotechar: str = '"'
    task_id: Optional[str] = None

    @field_validator("drive_folder_id")
    @classmethod
    def extract_drive_id(cls, v: Optional[str]) -> Optional[str]:
        if v and "/folders/" in v:
            match = re.search(r"/folders/([a-zA-Z0-9-_]+)", v)
            return match.group(1) if match else v
        return v


class ExportFileResult(BaseModel):
    site: str
    path: str
    folder_path: str
    rows: int


class ExportResult(BaseModel):
    status: Literal["success"]
    message: str
    files: List[ExportFileResult]
    drive_success: int


class PreviewResult(BaseModel):
    preview: str


class PreviewSitesResult(BaseModel):
    sites: List[str]
    sheets: List[str]
    columns: List[str]


class OpenFileRequest(BaseModel):
    path: Optional[str] = None


class OpenFileResult(BaseModel):
    status: Literal["success", "error"]

File: frontend/src/types/export.ts
import type { KoboForm } from './index';

export type ExportFormat = 'xlsx' | 'csv';
export type CsvSeparator = ';' | ',';
export type CsvEncoding = 'utf-8-sig' | 'utf-8' | 'windows-1252';

export interface SheetStructure {
  name: string;
  columns: string[];
}

export interface FormStructure {
  sheets: SheetStructure[];
}

export type AccountFormsMap = Record<number, KoboForm[]>;

export interface AccountFormPair {
  account_id: number;
  form_uid: string;
}

export interface SubmissionItem {
  id: string;
  submission_time?: string;
  site?: string;
  label?: string;
}

export interface ExportRequest {
  account_forms: AccountFormPair[];
  form_name: string;
  pivot_column?: string;
  selected_columns?: string[];
  selected_sheets?: string[];
  filter_sites?: string[];
  filter_submission_ids?: string[];
  drive_folder_id?: string;
  export_format: ExportFormat;
  csv_separator: CsvSeparator;
  csv_encoding: CsvEncoding;
  csv_quotechar: string;
  task_id?: string;
}

export interface PreviewRequest {
  account_forms: AccountFormPair[];
  form_name: string;
  csv_separator: CsvSeparator;
  csv_encoding: CsvEncoding;
  csv_quotechar: string;
  selected_sheets: string[];
}

export interface ExportFileResult {
  site: string;
  path: string;
  folder_path: string;
  rows: number;
  drive_link?: string;
  server_file_exists?: boolean;
}

export interface ExportResult {
  status: string;
  message: string;
  files: ExportFileResult[];
  directory?: string;
  drive_success?: number;
  drive_errors?: string[];
}

export interface SessionExportItem {
  id: string | number;
  timestamp: string;
  formName: string;
  format: ExportFormat;
  status: 'success' | 'cancelled' | 'error';
  message: string;
  files: ExportFileResult[];
  directory?: string;
  driveSuccess?: number;
  driveErrors?: string[];
  errorMessage?: string;
  createdAt?: string;
}

export interface PreviewResult {
  preview: string;
}

export interface PreviewSitesResult {
  sites?: string[];
  sheets?: string[];
  columns?: unknown[];
  submissions?: SubmissionItem[];
}

export interface CsvPrefs {
  format?: ExportFormat;
  sep?: CsvSeparator;
  enc?: CsvEncoding;
  quote?: string;
}

export interface GoogleStatus {
  connected: boolean;
}

